### DISCUSSION TITLE
Workflows to update the default list of opaque ports for CNCF projects, like NATS

### DISCUSSION THREAD

**Discussion by guhuajun** [linkerd/linkerd2#8438] [Ideas]
Greetings,

I found my NATS traffic was blocked by linkerd as _a protocol detection timeout_, I have documented my findings in one discussion of nats-py project, https://github.com/nats-io/nats.py/discussions/307

For specific Linkerd version, it's stable-2.11.0.

From linkerd project perspective, will it be possible to provide an official workflow to update to update the default list of opaque ports for CNCF projects. Or will linkerd project setup a formal validation process for CNCF projects?

Thanks,
Greg

**Comment by olix0r** [comment]
We'll consider changes to the default list of ports on a case-by-case basis. If you think NATS ports should be added to the default list, please open an issue!

> From linkerd project perspective, will it be possible to provide an official workflow to update to update the default list of opaque ports for CNCF projects. Or will linkerd project setup a formal validation process for CNCF projects?

If community members volunteer to audit the many CNCF projects to identify protocols that are incompatible with protocol detection, we'll welcome input about what changes should be made to Linkerd's defaults. But the core team doesn't have bandwidth to do this work.

**Reply by guhuajun** [reply]
Yes, bandwidth is always the problem. Would core team provide some guidelines (or rule of thumbs) to volunteers for auditing purpose? 
Appreciated core team's effort to enusre tcp proxying is possible, let's moving forward. To be honst, Linkerd is my favorite service mesh implementation so far.

Together for a shared future.

**Comment by wallyqs** [comment]
Hi I'm one of the NATS maintainers and just wanted to chime-in that would be great to collaborate to have support by default to avoid this issue. Would this be the appropriate list to add the NATS protocol? https://github.com/linkerd/linkerd2/blob/7ac79b8c086d177fafcd2bf01f6f6fb2e3949e98/charts/linkerd-control-plane/values.yaml#L161

**Reply by olix0r** [reply]
Sorry, I missed this message. Yes, that would be the right place to add mark NATS ports as opaque.

**Comment by fullykubed** [comment]
@wallyqs @olix0r Hate to revive an old thread, but I am experiencing this same issue even with opaque ports explicitly added on the NATS nodes.

I've documented my findings [here](https://github.com/linkerd/linkerd2/issues/13364), but I am at a loss on how to further debug or what might be causing this odd behavior under the hood.

**Comment by PROxZIMA** [comment]
@fullykubed @guhuajun I ran into the same problem. After a quick discussion with Linkerd support, I was able to resolve the timeout/connectivity issue. You can also check the comments in this PR for more details: https://github.com/PROxZIMA/nats-mtls/pull/1.
